Sources/NIOSSH/Connection State Machine/SSHConnectionStateMachine.swift

```diff
@@ -59,7 +59,10 @@ struct SSHConnectionStateMachine {
 
     /// The state of this state machine.
     private var state: State
-
+    
+    /// The username which authenticated to the server, if that has happened
+    private(set) var username: String? = nil
```

Comment:
Rather than store this as a stored property here, we should make it a computed property that calls into a computed property on `State`. That computed property can then switch over the state and load it from the underlying state data structure.

Reply:
This almost worked, but ran into trouble because an escaping closure cannot keep a mutating self reference. There just isn't any information to tie the successful conclusion of a authentication delegate to the username without tying it up in a closure or changing the delegate API which would be bad.

The next proposed solution introduces a group of mutable attributes on the SSHConnectionStateMachine which are in a class object, so can be handed around and updated as needed.